ReactWindows/ReactNative/UIManager/UIViewOperationQueueCollection.cs

```diff
@@ -0,0 +1,606 @@
+                        {
+                            _mainUiViewOperationsQueueInstance = null;
+                        }
+
+                        break;
+                    }
+                }
+            }
+
+        }
+
+        /// <summary>
+        /// Enqueues an operation to dispatch a command.
+        /// </summary>
+        /// <param name="tag">The view tag.</param>
+        /// <param name="commandId">The command identifier.</param>
+        /// <param name="commandArgs">The command arguments.</param>
+        public void EnqueueDispatchCommand(int tag, int commandId, JArray commandArgs)
+        {
+            // Called on layout manager thread
+
+            UIViewOperationQueueInstance queue;
+            if (!_reactTagToOperationQueue.TryGetValue(tag, out queue))
+            {
+                throw new InvalidOperationException("No queue for tag " + tag);
+            }
+
+            queue.EnqueueDispatchCommand(tag, commandId, commandArgs);
+        }
+
+        /// <summary>
+        /// Enqueues an operation to update the extra data for a view.
+        /// </summary>
+        /// <param name="reactTag">The view tag.</param>
+        /// <param name="extraData">The extra data.</param>
+        public void EnqueueUpdateExtraData(int reactTag, object extraData)
+        {
+            // Called on layout manager thread
+
+            UIViewOperationQueueInstance queue;
+            if (!_reactTagToOperationQueue.TryGetValue(reactTag, out queue))
+            {
+                throw new InvalidOperationException("No queue for tag " + reactTag);
+            }
+
+            queue.EnqueueUpdateExtraData(reactTag, extraData);
+        }
+
+        /// <summary>
+        /// Enqueues an operation to show a popup menu.
+        /// </summary>
+        /// <param name="tag">The view tag.</param>
+        /// <param name="items">The menu items.</param>
+        /// <param name="error">Called on error.</param>
+        /// <param name="success">Called on success.</param>
+        public void EnqueueShowPopupMenu(int tag, string[] items, ICallback error, ICallback success)
+        {
+            // Called on layout manager thread
+
+            UIViewOperationQueueInstance queue;
+            if (!_reactTagToOperationQueue.TryGetValue(tag, out queue))
+            {
+                throw new InvalidOperationException("No queue for tag " + tag);
+            }
+
+            queue.EnqueueShowPopupMenu(tag, items, error, success);
+        }
+
+        /// <summary>
+        /// Enqueues a operation to execute a UIBlock.
+        /// </summary>
+        /// <param name="block">The UI block.</param>
+        public void EnqueueUIBlock(IUIBlock block)
+        {
+            // Called on layout manager thread
+
+            // Native animation module is synchronized to the main dispatcher thread.
+            // Always forward to the main queue
+            MainUIViewOperationQueue.EnqueueUIBlock(block);
+        }
```

Comment:
Does this mean native animation module is not supported on secondary windows?

Reply:
Oh no, it is. It's just that animation module is a single threaded engine with all code executed in the main dispatcher thread. This includes:
- commands to set up "nodes" (like animated values, interpolations, attached views, etc.) - this is done through these UIBlock objects
- the actual computations and handling of the views on screen (engine hooks to orchestrator and calls SynchronouslyUpdateViewOnDispatcherThread below).

I had to choose between making the engine some multithreaded beast and the poor man's solution we have in the PR. I think the latter is much less code disturbing.
We can't create per thread instances as with the other cases because "nodes" are created prior to attaching views, and in theory some animated value can drive views created on separate threads anyway.

So everything remains on main thread, and it's the SynchronouslyUpdateViewOnDispatcherThread below that breaks the smooth flow and dispatches to the "right" thread. Presumably this makes secondary windows second class citizens with regards to animation, but I didn't see any rough UI during my testing.

---
In reply to: [165960540](https://github.com/Microsoft/react-native-windows/pull/1615#discussion_r165960540) [](ancestors = 165960540)